### ISSUE TITLE
.mgz / .mgh files created by mrtrix tools are not correctly read by FreeSurfer's tools

### ISSUE THREAD

**Issue by ansk** [MRtrix3/mrtrix3#952]
Dear experts,
I found that the FreeSurfer tools are segfaulting when reading the `.mgz / .mgh` files created by mrtrix tools. I tested this on FreeSurfer 5.3 as well as on dev version of FreeSurfer from 2/2017. I am using most recent build of mrtrix on master branch: 0.3.15-522-gbc67f1a1

Example: `orig.mgz` is file created by freeSurfer.
```
mrconvert orig.mgz orig_mrtrix.mgz

mrinfo orig_mrtrix.mgz 
************************************************
Image:               "orig_mrtrix.mgz"
************************************************
  Dimensions:        256 x 256 x 256
  Voxel size:        1 x 1 x 1
  Data strides:      [ -1 3 -2 ]
  Format:            MGZ (compressed MGH)
  Data type:         unsigned 8 bit integer
  Intensity scaling: offset = 0, multiplier = 1
  Transform:                    1   -1.11e-16   4.657e-10      -131.2
                        5.551e-17           1   7.276e-12      -114.8
                       -4.657e-10  -7.276e-12           1      -101.7

mri_info orig_mrtrix.mgz 
Floating point exception

mri_convert orig_mrtrix.mgz orig_mrtrix2.mgz

mri_convert.bin orig_mrtrix.mgz orig_mrtrix2.mgz 
$Id: mri_convert.c,v 1.226 2016/02/26 16:15:24 mreuter Exp $
reading from orig_mrtrix.mgz...
/hydra-db/hydra_io/vypocty/skoch/bin/freesurfer_dev_20170208/bin/mri_convert: line 3: 11714 Segmentation fault      mri_convert.bin "$@"

mri_binarize --i orig_mrtrix.mgz --min 1 --max 255 --o orig_mrtrix_binarized.mgz --count junk.txt

$Id: mri_binarize.c,v 1.43 2016/06/09 20:46:21 greve Exp $
cwd /hydradb/hydra_io/vypocty/freeSurfer/v6.1beta/CANEF/A_in_progress/skoch/ESO_P00007_20100611_1217_1_base/mri
cmdline mri_binarize.bin --i orig_mrtrix.mgz --min 1 --max 255 --o orig_mrtrix_binarized.mgz --count junk.txt 
sysname  Linux
hostname sup1.ad.nudz.cz
machine  x86_64
user     skoch

input      orig_mrtrix.mgz
frame      0
nErode3d   0
nErode2d   0
output     orig_mrtrix_binarized.mgz
Binarizing based on threshold
min        1
max        255
binval        1
binvalnot     0
/hydra-db/hydra_io/vypocty/skoch/bin/freesurfer_dev_20170208/bin/mri_binarize: line 3: 11922 Segmentation fault      mri_binarize.bin "$@"
```

**Comment by jdtournier** [comment]
> Floating point exception

That would suggest an unhandled divide-by-zero, square root of negative number, or overflow/underflow - most likely the former. Any chance you could find a (small) `mgz` file that works in FreeSurfer, but causes a crash if converted through `mrconvert`? i.e. a file where this would apply:
```bash
$ mri_info image.mgz                               # ok
$ mrconvert image.mgz image_converted.mgz          # ok
$ mri_info image_converted.mgz                     # crashes
```
We could then investigate what the difference between the original and converted files might be that causes the crash.

**Comment by ansk** [comment]
@jdtournier ,
see the attached files processed by the commands you specified.

[images.tar.gz](https://github.com/MRtrix3/mrtrix3/files/892543/images.tar.gz)

**Comment by Lestropie** [comment]
OK, thanks for the report.

Turns out that I was not sufficiently conservative in my `.mgh` / `.mgz` image outputs (and probably didn't test them sufficiently since I only ever use the read functionality...). The output images were being created with whatever byte-endianness was requested in the image header; this endianness would also be used when writing the header contents, as is done with NIfTI. It would seem that `mri_info` is failing catastrophically when reading a little-endian `.mgh` / `.mgz` image, which is surprising given the format includes a magic number at the start and therefore checking that magic number should fail if the reader code is not capable of detecting the swapped endianness.

Branch `mgh_mgz_write_fix` (Pull Request #956) for me produces byte-to-byte identical headers in output `mgh` files between `mri_convert` and `mrconvert`. Give it a test yourself if you want.

I did however just notice that there is a one-byte offset in the location of the "other" section between the outputs of `mri_convert` and `mrconvert`. *MRtrix3* is, as far as I can tell, [doing as it should](https://github.com/MRtrix3/mrtrix3/blob/master/lib/formats/mgh.cpp#L112-L114); so either I've overlooked something or `mri_info` is getting it wrong (the former being *faaar* more likely). Either way it requires a closer look.

**Comment by ansk** [comment]
Thank you for your time. I had tested the `mrconvert` and `mrcalc` in branch `mgh_mgz_write_fix`. This time the `mri_info` on `image_converted.mgz` does not crash, but it appears that the "metadata" is lost or not properly read both in `mrinfo` and `mri_info`:
```
$ release/bin/mrconvert image.mgz image_converted.mgz
```
```
$ release/bin/mrinfo image.mgz
************************************************
Image:               "image.mgz"
************************************************
  Dimensions:        256 x 256 x 256
  Voxel size:        1 x 1 x 1
  Data strides:      [ -1 3 -2 ]
  Format:            MGZ (compressed MGH)
  Data type:         unsigned 8 bit integer
  Intensity scaling: offset = 0, multiplier = 1
  Transform:                    1   -1.11e-16   4.657e-10      -131.2
                        5.551e-17           1   7.276e-12      -114.8
                       -4.657e-10  -7.276e-12           1      -101.7
  comments:          TR: 2300ms
                     Flip: 0.174532935deg
                     TE: 4.63000011ms
                     TI: 900ms
                     �/hydradb/hydra_io/vypocty/freeSurfer/v6.1beta/CANEF/A_in_progress/skoch/ESO_P00007_20100611_1217_1_base/mri/transforms/talairach.xfm!@AutoAlign   1.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   1.000000)UNKNOWN+*2P!@AutoAlign   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000Kmri_convert /tmpcount/freeSurfer/A_input/skochv/export_20_4_2016/ESO_P00007_20100611_1217_1/0006_t1_sag_mpr_3Diso__1mm/MR.1.3.12.2.1107.5.2.32.35226.20100611124804148709128 /tmpcount/freeSurfer/A_results/A_in_progress/skochv/pat/ESO_P00007_20100611_1217_1/mri/orig/001.mgz ProgramVersion: $Name: stable5 $  TimeStamp: 2016/04/21-13:32:39-GMT  BuildTimeStamp: May 13 2013 16:24:28  CVS: $Id: mri_convert.c,v 1.179.2.7 2012/09/05 21:55:16 mreuter Exp $  User: skochv  Machine: matpcp.ad.nudz.cz  Platform: Linux  PlatformVersion: 3.2.0-4-amd64  CompilerName: GCC  CompilerVersion: 40400  )mri_convert.bin /hydradb/hydra_io/vypocty/freeSurfer/v6.1beta/CANEF/A_in_progress/skoch/ESO_P00007_20100611_1217_1/mri/rawavg.mgz /hydradb/hydra_io/vypocty/freeSurfer/v6.1beta/CANEF/A_in_progress/skoch/ESO_P00007_20100611_1217_1/mri/orig.mgz --conform ProgramVersion: $Name:  $  TimeStamp: 2017/02/13-09:32:22-GMT  BuildTimeStamp: Feb  6 2017 17:11:52  CVS: $Id: mri_convert.c,v 1.226 2016/02/26 16:15:24 mreuter Exp $  User: skoch  Machine: sup4.ad.nudz.cz  Platform: Linux  PlatformVersion: 3.16.0-4-amd64  CompilerName: GCC  CompilerVersion: 40400  rmri_normalize -g 1 -mprage -W ctrl_vol.mgz bias_vol.mgz nu.mgz T1.mgz ProgramVersion: $Name:  $  TimeStamp: 2017/04/03-21:52:47-GMT  BuildTimeStamp: Feb  6 2017 17:11:52  CVS: $Id: mri_normalize.c,v 1.92 2017/01/27 22:31:34 ah221 Exp $  User: skoch  Machine: sup2.ad.nudz.cz  Platform: Linux  PlatformVersion: 3.16.0-4-amd64  CompilerName: GCC  CompilerVersion: 40400

```

```
$ release/bin/mrinfo image_converted.mgz
************************************************
Image:               "image_converted.mgz"
************************************************
  Dimensions:        256 x 256 x 256
  Voxel size:        1 x 1 x 1
  Data strides:      [ -1 3 -2 ]
  Format:            MGZ (compressed MGH)
  Data type:         unsigned 8 bit integer
  Intensity scaling: offset = 0, multiplier = 1
  Transform:                    1   -1.11e-16   4.657e-10      -131.2
                        5.551e-17           1   7.276e-12      -114.8
                       -4.657e-10  -7.276e-12           1      -101.7

```
Note zero values in acquisition parameters and missing talairach xfm info in mri_info:
```
$ mri_info image.mgz 
Volume information for image.mgz
          type: MGH
    dimensions: 256 x 256 x 256
   voxel sizes: 1.000000, 1.000000, 1.000000
          type: UCHAR (0)
           fov: 256.000
           dof: 0
        xstart: -128.0, xend: 128.0
        ystart: -128.0, yend: 128.0
        zstart: -128.0, zend: 128.0
            TR: 2300.00 msec, TE: 4.63 msec, TI: 900.00 msec, flip angle: 10.00 degrees
       nframes: 1
       PhEncDir: UNKNOWN
       FieldStrength: 0.000000
ras xform present
    xform info: x_r =  -1.0000, y_r =  -0.0000, z_r =  -0.0000, c_r =    -4.2219
              : x_a =  -0.0000, y_a =  -0.0000, z_a =   1.0000, c_a =    13.1834
              : x_s =   0.0000, y_s =  -1.0000, z_s =  -0.0000, c_s =    25.3142

talairach xfm : /hydradb/hydra_io/vypocty/freeSurfer/v6.1beta/CANEF/A_in_progress/skoch/ESO_P00007_20100611_1217_1_base/mri/transforms/talairach.xfm
Orientation   : LIA
Primary Slice Direction: coronal

voxel to ras transform:
               -1.0000  -0.0000  -0.0000   123.7781
               -0.0000  -0.0000   1.0000  -114.8166
                0.0000  -1.0000  -0.0000   153.3142
                0.0000   0.0000   0.0000     1.0000

voxel-to-ras determinant -1

ras to voxel transform:
               -1.0000  -0.0000   0.0000   123.7781
               -0.0000  -0.0000  -1.0000   153.3142
               -0.0000   1.0000  -0.0000   114.8166
               -0.0000  -0.0000  -0.0000     1.0000
```
```
$ mri_info image_converted.mgz 
Volume information for image_converted.mgz
          type: MGH
    dimensions: 256 x 256 x 256
   voxel sizes: 1.000000, 1.000000, 1.000000
          type: UCHAR (0)
           fov: 256.000
           dof: 0
        xstart: -128.0, xend: 128.0
        ystart: -128.0, yend: 128.0
        zstart: -128.0, zend: 128.0
            TR: 0.00 msec, TE: 0.00 msec, TI: 0.00 msec, flip angle: 0.00 degrees
       nframes: 1
       PhEncDir: UNKNOWN
       FieldStrength: 0.000000
ras xform present
    xform info: x_r =  -1.0000, y_r =  -0.0000, z_r =  -0.0000, c_r =    -4.2219
              : x_a =  -0.0000, y_a =  -0.0000, z_a =   1.0000, c_a =    13.1834
              : x_s =   0.0000, y_s =  -1.0000, z_s =  -0.0000, c_s =    25.3142

talairach xfm : 
Orientation   : LIA
Primary Slice Direction: coronal

voxel to ras transform:
               -1.0000  -0.0000  -0.0000   123.7781
               -0.0000  -0.0000   1.0000  -114.8166
                0.0000  -1.0000  -0.0000   153.3142
                0.0000   0.0000   0.0000     1.0000

voxel-to-ras determinant -1

ras to voxel transform:
               -1.0000  -0.0000   0.0000   123.7781
               -0.0000  -0.0000  -1.0000   153.3142
               -0.0000   1.0000  -0.0000   114.8166
               -0.0000  -0.0000  -0.0000     1.0000

```

Is it caused by that one byte offset you are referring to?

**Comment by Lestropie** [comment]
Surprisingly: no it's not.

When writing specifically to `.mgz`, *MRtrix3* won't include any metadata over and above what goes into the fixed-size image file header. The [`mgh` format](https://surfer.nmr.mgh.harvard.edu/fswiki/FsTutorial/MghFormat) stores such additional metadata *after* the image data, which is fundamentally incompatible with how compressed image formats are currently supported in the *MRtrix3* code. So converting to `.mgh` will preserve that information, but `.mgz` will not.

Although I'm sure in principle support for this could be added, it would require a huge amount of gymnastics in order to support this one use case. Personally I don't think it's worth the expense, given one can simply gzip a `.mgh` and rename it to `.mgz` after the fact.

-----------

The one-byte offset isn't what I thought it was: The metadata areas for `mri_convert` and `mrconvert` both *start* at the same location, but the example file you provided contains massive null-filled blocks in between comments within this region, and it just so happens that somewhere a single null byte gets inconspicuously dropped by `mrconvert`. The current [import code](https://github.com/MRtrix3/mrtrix3/blob/master/lib/file/mgh_utils.cpp#L127-L128) isn't as robust as it could be... Nevertheless, this one-byte difference doesn't actually trip up either software.

**Comment by Lestropie** [comment]
Ugh. Okay, this metadata is annoying...

I tried formatting the incoming metadata to get cleaner comment fields with no control characters or null space or anything. That works (see 5aaaa23). But if such "clean" comments are written to a `.mgh` using `mrconvert`, `mri_info` fails catastrophically. Conversely, if I *retain* all of the junk that comes out of the metadata as-is, then write it as-is to a `.mgh` using `mrconvert`, `mri_info` is happy (though I suspect fragile, depending on whether these comments are modified in any way).

So the metadata field "[tags](https://surfer.nmr.mgh.harvard.edu/fswiki/FsTutorial/MghFormat)" is not actually just null-terminated strings, but seems to contain some higher-order information, at least part of which is used to determine memory allocation in `mri_info`.

As an example, here's what I get out of `mrinfo image.mgz -property comments`, with up-to-date code on `mgh_mgz_write_fix` branch:

```
TR: 2300ms
Flip: 10deg
TE: 4.63ms
TI: 900ms
/hydradb/hydra_io/vypocty/freeSurfer/v6.1beta/CANEF/A_in_progress/skoch/ESO_P00007_20100611_1217_1_base/mri/transforms/talairach.xfm
!
@AutoAlign   1.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   1.000000
)
UNKNOWN
+
*
2P
!
@AutoAlign   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000
Kmri_convert /tmpcount/freeSurfer/A_input/skochv/export_20_4_2016/ESO_P00007_20100611_1217_1/0006_t1_sag_mpr_3Diso__1mm/MR.1.3.12.2.1107.5.2.32.35226.20100611124804148709128 /tmpcount/freeSurfer/A_results/A_in_progress/skochv/pat/ESO_P00007_20100611_1217_1/mri/orig/001.mgz ProgramVersion: $Name: stable5 $  TimeStamp: 2016/04/21-13:32:39-GMT  BuildTimeStamp: May 13 2013 16:24:28  CVS: $Id: mri_convert.c,v 1.179.2.7 2012/09/05 21:55:16 mreuter Exp $  User: skochv  Machine: matpcp.ad.nudz.cz  Platform: Linux  PlatformVersion: 3.2.0-4-amd64  CompilerName: GCC  CompilerVersion: 40400  
)mri_convert.bin /hydradb/hydra_io/vypocty/freeSurfer/v6.1beta/CANEF/A_in_progress/skoch/ESO_P00007_20100611_1217_1/mri/rawavg.mgz /hydradb/hydra_io/vypocty/freeSurfer/v6.1beta/CANEF/A_in_progress/skoch/ESO_P00007_20100611_1217_1/mri/orig.mgz --conform ProgramVersion: $Name:  $  TimeStamp: 2017/02/13-09:32:22-GMT  BuildTimeStamp: Feb  6 2017 17:11:52  CVS: $Id: mri_convert.c,v 1.226 2016/02/26 16:15:24 mreuter Exp $  User: skoch  Machine: sup4.ad.nudz.cz  Platform: Linux  PlatformVersion: 3.16.0-4-amd64  CompilerName: GCC  CompilerVersion: 40400  
rmri_normalize -g 1 -mprage -W ctrl_vol.mgz bias_vol.mgz nu.mgz T1.mgz ProgramVersion: $Name:  $  TimeStamp: 2017/04/03-21:52:47-GMT  BuildTimeStamp: Feb  6 2017 17:11:52  CVS: $Id: mri_normalize.c,v 1.92 2017/01/27 22:31:34 ah221 Exp $  User: skoch  Machine: sup2.ad.nudz.cz  Platform: Linux  PlatformVersion: 3.16.0-4-amd64  CompilerName: GCC  CompilerVersion: 40400 
```

(Yes, that's *after* removing non-printable characters)

What I've chosen to do is:

- When reading from a `.mgh` / `.mgz`, format the incoming metadata as best as I can.

- When writing to `.mgh` / `.mgz`, *omit* all text metadata.

"Be conservative in what you send, be liberal in what you accept": Grab what we can from the MGH metadata, but don't write metadata that (even though it technically conforms to their format documentation) may break FreeSurfer. Unless someone is able to provide information on the MGH metadata formatting over and above what's on [this page](https://surfer.nmr.mgh.harvard.edu/fswiki/FsTutorial/MghFormat).

**Comment by Lestropie** [comment]
Should be solved by #970. Let us know if there are still ongoing issues.
